builtin/credential/aws/path_role.go

```diff
@@ -241,6 +282,16 @@ func (b *backend) nonLockedAWSRole(s logical.Storage, roleName string) (*awsRole
 		}
 	}
 
+	// Check if there was no pre-existing AuthType set (from older versions)
```

Comment:
This is a wrong place to do this upgrade. `nonLockedAWSRole` will be invoked with a read lock and not a write lock. Also, the block above (which are not your changes) seems like a bug which is doing exactly the same thing.

If you don't mind, can you please move both of these blocks to `lockedAWSRole` method? In `lockedAWSRole`, after `nonLockedAWSRole` returns, we should invoke `lockedSetAWSRole` to perform a race free upgrade.